File: API/Services/MovieService .cs
using API.DTOs;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;

namespace API.Services
{
    public class MovieService
    {
        private readonly HttpClient _httpClient;

        public MovieService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        public async Task<List<MovieDto>> GetMoviesAsync()
        {
            var response = await _httpClient.GetAsync("https://movie.pequla.com/api/movie");
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException($"Greška u API pozivu: {response.StatusCode}");
            }

            var jsonString = await response.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<List<MovieDto>>(jsonString);
        }

        public async Task<List<MovieDto>> GetFilteredSortedSearchedMovies(
            string searchTerm, 
            string genres, 
            string actors, 
            string director, 
            int? minDuration, 
            int? maxDuration, 
            string startDate, 
            string endDate, 
            string orderBy)
        {
            var movies = await GetMoviesAsync();

            // Pretraga po terminu
            if (!string.IsNullOrEmpty(searchTerm))
            {
                movies = movies.Where(m =>
                    m.Title.ToLower().Contains(searchTerm.ToLower()) ||
                    m.Description.ToLower().Contains(searchTerm.ToLower()) ||
                    m.ShortDescription.ToLower().Contains(searchTerm.ToLower())
                ).ToList();
            }

           if (!string.IsNullOrEmpty(genres))
{
            var genreTerms = genres.ToLower().Split(",").Select(g => g.Trim()).ToList();
            movies = movies.Where(m =>
                m.MovieGenres.Any(g =>
                    genreTerms.Any(term => g.Genre.Name.ToLower().Contains(term))
                )
            ).ToList();
        }

            // Filtriranje po glumcima
            if (!string.IsNullOrEmpty(actors))
            {
                var actorList = actors.ToLower().Split(",").ToList();
                movies = movies.Where(m =>
                    m.MovieActors.Any(a => actorList.Contains(a.Actor.Name.ToLower()))
                ).ToList();
            }

            // Filtriranje po reditelju
            if (!string.IsNullOrEmpty(director))
            {
                movies = movies.Where(m =>
                    m.Director.Name.ToLower().Contains(director.ToLower())
                ).ToList();
            }

            // Filtriranje po minimalnom trajanju
            if (minDuration.HasValue)
            {
                movies = movies.Where(m => m.RunTime >= minDuration.Value).ToList();
            }

            // Filtriranje po maksimalnom trajanju
            if (maxDuration.HasValue)
            {
                movies = movies.Where(m => m.RunTime <= maxDuration.Value).ToList();
            }

            // Filtriranje po datumu početka
            if (!string.IsNullOrEmpty(startDate))
            {
                var startDateParsed = DateTime.Parse(startDate);
                movies = movies.Where(m => DateTime.Parse(m.StartDate) >= startDateParsed).ToList();
            }

            // Filtriranje po datumu kraja
            if (!string.IsNullOrEmpty(endDate))
            {
                var endDateParsed = DateTime.Parse(endDate);
                movies = movies.Where(m => DateTime.Parse(m.StartDate) <= endDateParsed).ToList();
            }

            // Sortiranje
            if (!string.IsNullOrEmpty(orderBy))
            {
                switch (orderBy)
                {
                    case "  ":
                        movies = movies.OrderBy(m => m.RunTime).ToList();
                        break;
                    case "durationDesc":
                        movies = movies.OrderByDescending(m => m.RunTime).ToList();
                        break;
                    case "title":
                        movies = movies.OrderBy(m => m.Title).ToList();
                        break;
                    case "titleDesc":
                        movies = movies.OrderByDescending(m => m.Title).ToList();
                        break;
                    default:
                        movies = movies.OrderBy(m => m.Title).ToList();
                        break;
                }
            }

            return movies;
        }

        // Dodajemo GetGenresAsync metodu
        public async Task<List<string>> GetGenresAsync()
        {
            var response = await _httpClient.GetAsync("https://movie.pequla.com/api/genres");
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException($"Greška u API pozivu: {response.StatusCode}");
            }

            var jsonString = await response.Content.ReadAsStringAsync();
            var genres = JsonConvert.DeserializeObject<List<string>>(jsonString);
            return genres;
        }


        public async Task<MovieDto> GetMovieByIdAsync(int id)
        {
            var response = await _httpClient.GetAsync($"https://movie.pequla.com/api/movie/{id}");
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException($"Greška u API pozivu: {response.StatusCode}");
            }

            var jsonString = await response.Content.ReadAsStringAsync();
            var movie = JsonConvert.DeserializeObject<MovieDto>(jsonString);
            return movie;
        }
    }
}

File: client-app/src/app/core/services/movie.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Movie } from '../../model/movie.model';

@Injectable({
  providedIn: 'root'
})
export class MovieService {
  private apiUrl = 'http://localhost:5000/api/Movie';
  private http = inject(HttpClient);

  getMovies(): Observable<Movie[]> {
    return this.http.get<Movie[]>(this.apiUrl);
  }

  getMovieById(id: number): Observable<Movie> {
    return this.http.get<Movie>(`${this.apiUrl}/${id}`);
  }
  getGenresAsString(genres: any[]): string {
    console.log(genres)
    return genres.map(g => g.genre.name).join(', ');
  }
  
  // Add this method to handle filtered movies
  getFilteredMovies(filters: any): Observable<Movie[]> {
    let params = new HttpParams();

    // Append the filters to the HTTP params
    if (filters.searchTerm) {
      params = params.append('searchTerm', filters.searchTerm);
    }
    if (filters.genres) {
      params = params.append('genres', filters.genres);
    }
    if (filters.actors) {
      params = params.append('actors', filters.actors);
    }
    if (filters.director) {
      params = params.append('director', filters.director);
    }
    if (filters.minDuration) {
      params = params.append('minDuration', filters.minDuration.toString());
    }
    if (filters.maxDuration) {
      params = params.append('maxDuration', filters.maxDuration.toString());
    }
    if (filters.startDate) {
      params = params.append('startDate', filters.startDate);
    }
    if (filters.endDate) {
      params = params.append('endDate', filters.endDate);
    }
    if (filters.orderBy) {
      params = params.append('orderBy', filters.orderBy);
    }

    return this.http.get<Movie[]>(`${this.apiUrl}`, { params });
  }
}
